File: LEETCODE/LEETCODE-PYTHON/277.py
# 277. Find the Celebrity
# https://leetcode.com/problems/find-the-celebrity/description/

# The knows API is already defined for you.
# return a bool, whether a knows b
# def knows(a: int, b: int) -> bool:

class Solution:
    def findCelebrity(self, n: int) -> int:
        candidate=0

        for i in range(1,n):
            if(not knows(i,candidate)):
                candidate=i

        for j in range(0,n):
            if(j==candidate):
                continue
            
            if(not knows(j,candidate) or knows(candidate,j)):
                return -1

        return candidate

File: LEETCODE/LEETCODE-CPP/277.cpp
// 277. Find the Celebrity
// https://leetcode.com/problems/find-the-celebrity/description/

/* The knows API is defined for you.
      bool knows(int a, int b); */

class Solution
{
public:
    int findCelebrity(int n)
    {
        if (n <= 1)
        {
            return n;
        }

        int candidate = 0;

        for (int i = 1; i < n; i++)
        {
            if (!knows(i, candidate))
            {
                candidate = i;
            }
        }

        for (int j = 0; j < n; j++)
        {
            if (j == candidate)
            {
                continue;
            }

            if (!knows(j, candidate) || knows(candidate, j))
            {
                return -1;
            }
        }

        return candidate;
    }
};